### DISCUSSION TITLE
RSA key generation with native XS/C

### DISCUSSION THREAD

**Discussion by thebatt** [Moddable-OpenSource/moddable#862] [Q&A]
I am currently using the popular JSBN library for generating and encrypting/decrypting RSA messages, and I would like to implement something native on the XS side, for performance. I have adapted the RNG in JSBN to be able to generate deterministic keys from a passphrase, so a user passphrase can be used to decrypt messages encrypted by the server with a public key for the user, thus authenticating the user in a zero-trust manner. The server does the same to authenticate itself for the user's benefit, for mutual assurance of identity, having at no time any private keys going across the wire.

In the XS modules, is there an RSA key generator that can likewise be modified from the PRNG to the deterministic? I have not found an example script for RSA generation, encryption, and decryption.

**Comment by phoddie** [comment]
If you haven't already, do take a look at the [crypt directory](https://github.com/Moddable-OpenSource/moddable/tree/public/modules/crypt) in the Moddable SDK. There's quite a bit of fundamental crypto support there, more than enough to power our TLS implementation. Much of it is already optimized with native code. It is all built on a foundation of JavaScript's built-in `BigInt` type.

**Comment by thebatt** [comment]
I have implemented BigInt instead of the old JSBN library, and am now working on key generation, which, for RSA, does not appear in the TLS code, that is only for DH handshake, ECDSA and curves from what I can see.

**Reply by phoddie** [reply]
> I have implemented BigInt instead of the old JSBN library

That's great! Using `BigInt` makes the script smaller and more portable. It'll benefit from any optimization work to BigInt in the future.

> am now working on key generation, which, for RSA, does not appear in the TLS code

That may be the case.
